_blogposts/2025-04-05-introducing-unified-operators.mdx

```diff
@@ -0,0 +1,268 @@
+---
+author: rescript-team
+date: "2025-04-05"
+title: "Rethinking Operators: How ReScript Simplifies Arithmetic and Beyond"
+badge: roadmap
+description: |
+    Discover how unified operators in ReScript v12 simplify arithmetic, reduce syntax noise — plus, a glimpse into the future roadmap.
+---
+
+## Introduction
+
+For upcoming ReScript v12, we're upgrading common arithmetic operators to be "Unified Operators"
+
+This means that we can now use a single infix operator syntax for multiple numeric types, and even for string concatenation.
+
+```res
+let add_int = 1 + 2
+let add_float = 1.0 + 2.0
+let concat_string = "Hello" + ", World!"
+```
+
+[Try in the playground](https://rescript-lang.org/try?version=v12.0.0-alpha.10&module=esmodule&code=DYUwLgBAhgJjD6BLAdpAvBAjBA1BATAFCiSwIBmwA9lOlgHQAMuBTx4EAxlcp7fAGcwAJxQBzCBgBEACRDBqUllIA0EAOpVhwGAEIphIA)— it just works since `v12.0.0-alpha.5`. We don't need `+.` and `++` anymore. 🥳
+
+This post covers both the reasoning behind the change and what’s next on the roadmap. If you're interested in the implementation details, you’ll find them in [the pull request](https://github.com/rescript-lang/rescript/pull/7057).
+
+## Problems in operators
+
+Until v12, the operator syntax had a few notable problems.
+
+### Unwanted syntax gap
+
+Having different operators for each type is not familiar to JavaScript users, and not even allowing overloading can feel strange to most programmers.
+
+This is tricky in the real world. Because JavaScript's default number type is `float`, not `int`, ReScript users have to routinely deal with awkward syntax like `+.`, `-.`, `*.`, `%.`.
+
+Some operators are available only as functions. Instead of `<<` and `>>`, we use unfamiliar names like `lsl` and `asr`.
+
+### Infix explosion 💥
+
+ReScript has multiple "add operator" syntaxes for every primitive type.
+
+```res
+let add_int = 1 + 2
+let add_float = 1.0 +. 2.0
+let concat_string = "Hello" ++ ", World!"
+```
+
+We ran into the same issue again when we added `bigint` support.
+
+What other operator syntax could we introduce to add two bigint values? There were suggestions like `+,`, `+!`, `+n`, but the team never felt confident in any of them, so we just hid them inside the `BigInt` module definition instead of introducing new syntax.
+
+It was inconvenient because we had to shadow the definition every time.
+
+```res
+let add_bigint = {
+  open BigInt!
+  1n + 2n
+}
+```
+
+Every time we introduce a new primitive type (who knows!), we face the same issue across all arithmetic operators.
```

Comment:
Every time we introduce a new primitive type (who knows?), we run into the same issue with all arithmetic operators.